rustls-cert-gen/src/cert.rs

```diff
@@ -0,0 +1,433 @@
+		params.is_ca = IsCa::Ca(BasicConstraints::Unconstrained);
+		params.key_usages.push(KeyUsagePurpose::DigitalSignature);
+		params.key_usages.push(KeyUsagePurpose::KeyCertSign);
+		params.key_usages.push(KeyUsagePurpose::CrlSign);
+		Self { params }
+	}
+	pub fn country_name(mut self, country: &str) -> Self {
+		self.params
+			.distinguished_name
+			.push(DnType::CountryName, PrintableString(country.into()));
+		self
+	}
+	pub fn organization_name(mut self, name: &str) -> Self {
+		self.params
+			.distinguished_name
+			.push(DnType::OrganizationName, name);
+		self
+	}
+	pub fn build(self) -> Result<Ca, rcgen::Error> {
+		let cert = Certificate::from_params(self.params)?;
+		let cert = Ca { cert };
+		Ok(cert)
+	}
+}
+
+pub struct Ca {
+	cert: Certificate,
+}
+
+impl Ca {
+	/// Self-sign and serialize
+	pub fn serialize_pem(&self) -> Result<PemCertifiedKey, rcgen::Error> {
+		let cert_pem = self.cert.serialize_pem()?;
+		let key_pem = self.cert.serialize_private_key_pem();
+		Ok(PemCertifiedKey {
+			cert_pem,
+			private_key_pem: key_pem,
+		})
+	}
+	pub fn cert(&self) -> &Certificate {
+		&self.cert
+	}
+}
+
+/// End-entity [Certificate]
+pub struct EndEntity {
+	cert: Certificate,
+}
+
+impl EndEntity {
+	/// Sign with `self.signer` and serialize.
+	pub fn serialize_pem(&self, signer: &Certificate) -> Result<PemCertifiedKey, rcgen::Error> {
+		let cert_pem = self.cert.serialize_pem_with_signer(signer)?;
+		let key_pem = self.cert.serialize_private_key_pem();
+		Ok(PemCertifiedKey {
+			cert_pem,
+			private_key_pem: key_pem,
+		})
+	}
+}
+
+/// [CertificateParams] from which an [EndEntity] [Certificate] can be built
+pub struct EndEntityParams {
+	params: CertificateParams,
+}
+
+impl EndEntityParams {
+	/// Initialize `EndEntityParams`
+	pub fn new(mut params: CertificateParams) -> Self {
+		params.is_ca = IsCa::NoCa;
+		params.use_authority_key_identifier_extension = true;
+		params.key_usages.push(KeyUsagePurpose::DigitalSignature);
+		Self { params }
+	}
+	pub fn common_name(mut self, name: &str) -> Self {
+		self.params
+			.distinguished_name
+			.push(DnType::CommonName, name);
+		self
+	}
+	/// `SanTypes` that will be recorded as `subject_alt_names`
+	pub fn subject_alternative_names(mut self, sans: Vec<SanType>) -> Self {
+		for san in sans {
+			self.params.subject_alt_names.push(san);
+		}
+		self
+	}
+	/// Add ClientAuth to `extended_key_usages`.
+	pub fn client_auth(&mut self) -> &Self {
+		let usage = ExtendedKeyUsagePurpose::ClientAuth;
+		self.params.extended_key_usages.push(usage);
+		self
+	}
+	/// Add ServerAuth to `extended_key_usages`.
+	pub fn server_auth(&mut self) -> &Self {
+		let usage = ExtendedKeyUsagePurpose::ServerAuth;
+		self.params.extended_key_usages.push(usage);
+		self
+	}
+	pub fn build(self) -> Result<EndEntity, rcgen::Error> {
+		let cert = Certificate::from_params(self.params)?;
+		let cert = EndEntity { cert };
+		Ok(cert)
+	}
+}
+
+pub struct Signature<'a> {
```

Comment:
I think this type would be better expressed as a an enum named something like `KeypairAlgorithm` with variants `Ed25519`, `EcdsaP256` and `EcdsaP384`.

I think we should avoid "stringly-typed" inputs since they're hard for consumers to use. The `ed25519`, `ecdsa_p256` and `ecdsa_p384` functions on this type could be reconstituted as `KeypairAlgorithm::to_keypair` that converts to `rcgen::KeyPair` as appropriate.

Reply:
I agree. We can have the cli arg parser parse directly to an enum so that `to_keypair` can take a variant. I have implemented this in the latest commit.